docs/spec/v2/log.json

```diff
@@ -0,0 +1,839 @@
+                "null",
+                "string",
+                "integer"
+              ],
+              "maxLength": 1024
+            },
+            "username": {
+              "description": "Name of the user.",
+              "type": [
+                "null",
+                "string"
+              ],
+              "maxLength": 1024
+            }
+          }
+        }
+      }
+    },
+    "faas": {
+      "description": "FAAS holds fields related to Function as a Service events.",
+      "type": [
+        "null",
+        "object"
+      ],
+      "properties": {
+        "coldstart": {
+          "description": "Indicates whether a function invocation was a cold start or not.",
+          "type": [
+            "null",
+            "boolean"
+          ]
+        },
+        "execution": {
+          "description": "The request id of the function invocation.",
+          "type": [
+            "null",
+            "string"
+          ]
+        },
+        "id": {
+          "description": "A unique identifier of the invoked serverless function.",
+          "type": [
+            "null",
+            "string"
+          ]
+        },
+        "name": {
+          "description": "The lambda function name.",
+          "type": [
+            "null",
+            "string"
+          ]
+        },
+        "trigger": {
+          "description": "Trigger attributes.",
+          "type": [
+            "null",
+            "object"
+          ],
+          "properties": {
+            "request_id": {
+              "description": "The id of the origin trigger request.",
+              "type": [
+                "null",
+                "string"
+              ]
+            },
+            "type": {
+              "description": "The trigger type.",
+              "type": [
+                "null",
+                "string"
+              ]
+            }
+          }
+        },
+        "version": {
+          "description": "The lambda function version.",
+          "type": [
+            "null",
+            "string"
+          ]
+        }
+      }
+    },
+    "level": {
+      "description": "Level represents the severity of the recorded log.",
+      "type": [
+        "null",
+        "string"
+      ],
+      "maxLength": 1024
+    },
+    "logger_name": {
+      "description": "LoggerName holds the name of the used logger instance.",
+      "type": [
+        "null",
+        "string"
+      ],
+      "maxLength": 1024
+    },
+    "message": {
+      "description": "Message of the logged error. In case a parameterized message is captured, Message should contain the same information, but with any placeholders being replaced.",
+      "type": "string"
+    },
+    "severity": {
+      "description": "Severity represents the numreic severity of the logs as advertised by the source. Meaning behind severity is upto the implementer.",
+      "type": [
+        "null",
+        "integer"
+      ]
+    },
+    "span_id": {
```

Comment:
Note that unlike traces and spans, logs are expected to be already in ECS format, as we'll use our ECS logging libraries to convert log events to JSON. So it shouldn't be `span_id` but `span.id`. This also applies to `timestamp` -> `@timestamp`, `trace_id` -> `trace.id` and others.

I think it will end up being much easier to not even have a dedicated spec for `log`. Just forward everything agents send to Elasticserach.

Reply:
Having said that, the global metadata ofc needs to be converted to ECS.

What I had in mind is that APM Server would just send a doc like
```
{
  ... global metadata
  "message": "<whatever the agent has sent>"
}
```

And then use an ingest node pipeline to do a JSON decoding of the `message` field that merges the fields into the root of the doc.

Dynamic fields become an issue then. If we want to disallow dynamic fields in the first iteration, we can add `dynamic: false` to the mappings. However, we could also just add `dynamic: runtime`. This makes the fields searchable but avoids  (some) mapping issues related to dynamic fields and can save space and index time for fields that are not commonly used in searches.
Other mechanisms to avoid issues with dynamic fields are to enable `ignore_malformed` for the whole data stream and to enable `subobjects: true` on the root level once https://github.com/elastic/elasticsearch/issues/88934 has been resolved.

Reply:
> maybe something like raw.

Or we can put everything else under `labels` as suggested by Andrew in the slack conversation.

Reply:
IIANM, for Lambda all we get is a timestamp and a string for function logs. Is that right? If so then I agree with the suggestions above to just send `@timestamp` and `message`.

For the first cut, I would suggest we just store `message` as it is received. Then we can follow up with what Felix suggests, and attempt to decode the message as JSON into the root, and use `dynamic: runtime`.